docs/development/design/network-policy.md

```diff
@@ -0,0 +1,360 @@
+# Design Document: Network Policy
+
+## Introduction
+
+The goal of this design document is to outline the implementation of security groups. Security groups will contain a set of rules that specify allowed ports and protocols on the driver interface. This will be managed using nftables and initially support Linux devices only.
+
+## Phased Implementation
+
+### Phase I - Default Security Group for Organization
+
+- The default security group for an organization will be applied to all devices on startup.
+- A device can only have one policy group applied at any given time in phase I. This means each organization will have a security group. Users in the organization can CRUD the security group. AAlternatively we can scope it to only the organization owner that can modify the group.
+- When a new set of rules are applied, clear the appropriate chain and re-apply the rules. This can be more elegantly managed but introduces a great deal of complexity when rules are overlapping and may not be installed in the tables because another rule superceded it. See the section [Rule Deconfliction](#rule-deconfliction).
+
+### Phase II - User-owned Security Groups
+
+- Users can create their own security groups for more granular policies per device. The user owns the security group. It seems reasonable that other members of an org should be able to use that security group. This would be similar to how an ec2 security group gets applied to a node would be the same here with a patch.
+- A device can only have one policy group applied at any given time.
```

Comment:
This doesn't sound like a natural addition to phase 1. I don't think a user should be able to drop the org-admin level policy that was added in phase 1.

The RBAC I would expect in this phase is:

- Users CRUD their own policies.
- The org owner should be able to get details of user-level policies applied to each device in an organization.
- Org owners should NOT be able to CRUD or even list all policies defined by a user.

It's questionable whether all users should be able to view the user-level policies defined for other users' devices in the same organization. I'm tempted to err on the side of don't make it visible. It could be opened up more later.

Reply:
@chirino commented

> I'm not sure hiding device level policy from peers adds more security to the system. If two devices can peer with each other, I think you could find upsides to being able to inspect each other's policies.
> 
> For example if device A and B allow all outbound connections (the default) but deny all inbound the you could skip setting up peering between them.

That works for me. "if they can peer, viewing policy seems reasonable" sold me.

So forget my last paragraph and replace with: All org members should be able to view the user-level policies assigned to a device in the same organization.